src/Graphics/Color.cs

```diff
@@ -14,942 +14,1121 @@
 using System.Runtime.CompilerServices;
 using System.Text;
 using System.Numerics;
+using System.Runtime.InteropServices;
 
-namespace MoonWorks.Graphics
+namespace MoonWorks.Graphics;
+
+/// <summary>
+/// Describes a 32-bit packed color.
+/// </summary>
+[StructLayout(LayoutKind.Sequential, Size = 4, Pack = 4)]
+public record struct Color(byte R, byte G, byte B, byte A = 255)
 {
+	public Color(int r, int g, int b, int a = 255) : this(
+		(byte) int.Clamp(r, 0, 255),
+		(byte) int.Clamp(g, 0, 255),
+		(byte) int.Clamp(b, 0, 255),
+		(byte) int.Clamp(a, 0, 255)
+	) { }
+
+	/// <summary>
+	/// Constructs an RGBA color from scalars which represent red, green, blue and alpha values.
+	/// </summary>
+	/// <param name="r">Red component value from 0.0f to 1.0f.</param>
+	/// <param name="g">Green component value from 0.0f to 1.0f.</param>
+	/// <param name="b">Blue component value from 0.0f to 1.0f.</param>
+	/// <param name="alpha">Alpha component value from 0.0f to 1.0f.</param>
+	public Color(float r, float g, float b, float a) : this(
+		(byte) (System.Math.Clamp(r, 0, 1) * byte.MaxValue),
+		(byte) (System.Math.Clamp(g, 0, 1) * byte.MaxValue),
+		(byte) (System.Math.Clamp(b, 0, 1) * byte.MaxValue),
+		(byte) (System.Math.Clamp(a, 0, 1) * byte.MaxValue)
+	) { }
+
+	/// <summary>
+	/// Constructs an RGBA color from scalars which represent red, green, blue values. Alpha is assumed to be 1.
+	/// </summary>
+	/// <param name="r">Red component value from 0.0f to 1.0f.</param>
+	/// <param name="g">Green component value from 0.0f to 1.0f.</param>
+	/// <param name="b">Blue component value from 0.0f to 1.0f.</param>
+	public Color(float r, float g, float b) : this(
+		(byte) (System.Math.Clamp(r, 0, 1) * byte.MaxValue),
+		(byte) (System.Math.Clamp(g, 0, 1) * byte.MaxValue),
+		(byte) (System.Math.Clamp(b, 0, 1) * byte.MaxValue),
+		(byte) 255
+	) { }
+
+	public Color(Vector4 vector) : this(vector.X, vector.Y, vector.Z, vector.W) { }
+
 	/// <summary>
-	/// Describes a 32-bit packed color.
+	/// Gets packed RGBA value of this <see cref="Color"/>.
+	/// <para>NOTE: Because of endianness, if you format this value directly as a hex string, 
+	/// the order of bytes may be wrong.</para>
 	/// </summary>
-	public record struct Color(byte R, byte G, byte B, byte A = 255)
+	public readonly uint PackedValue() => Unsafe.BitCast<Color, uint>(this);
+
+	/// <summary>
+	/// Gets a <see cref="Color"/> representation for a packed RGBA value.
+	/// <para>NEVER plug in a value that wasn't generated by PackedValue(),
+	/// unless you're SURE you can deal with an endianness mismatch.</para>
+	/// <para>Example: it's NOT safe to take 0xF0F8FFFF from AliceBlue and plug it into here!</para>
+	/// </summary>
+	public static Color FromPacked(uint rgba) => Unsafe.BitCast<uint, Color>(rgba);
```

Comment:
I'm really not crazy about introducing an endianness gotcha here. If we're going to do this, we should just specify the endianness of the input.

Reply:
Well I ran around in circles for a bit, but the final result is a lot simpler: https://github.com/MoonsideGames/MoonWorks/pull/28/commits/2d36f3959a55967b06525247e3ac1cf4750ac166